lib/srv/desktop/rdp/rdpclient/src/rdpdr/scard.rs

```diff
@@ -1955,13 +1955,31 @@ impl WriteCache_Call {
     }
 }
 
+impl Encode for WriteCache_Call {
+    fn encode(&self) -> RdpResult<Message> {
+        let mut w = vec![];
+
+        w.extend(RPCEStreamHeader::new().encode()?);
+        RPCETypeHeader::new(0).encode(&mut w)?;
+
+        let mut index = 0;
+        encode_ptr(None, &mut index, &mut w)?; // _lookup_name_ptr
+        self.common.encode_ptr(&mut index, &mut w)?;
+
+        w.extend(encode_str_unicode(&self.lookup_name)?);
+        self.common.encode_value(&mut w)?;
+
+        Ok(w)
+    }
+}
+
 #[derive(Debug)]
 #[allow(non_camel_case_types, dead_code)]
-struct WriteCache_Common {
-    context: Context,
-    card_uuid: Vec<u8>,
-    freshness_counter: u32,
-    data: Vec<u8>,
+pub(crate) struct WriteCache_Common {
```

Comment:
Is this really needed to make everything public only for testing purposes? Could you for example move the test to the same package to avoid exporting?

Reply:
Yeah good point... This should all be done within `scard`. I'm going to close this PR (and its big stack of sibling PR's) and redo it.